Add largest-series-product exercise

Aksima opened the pull request:
This is intended to be a commit to the https://github.com/Aksima/groovy.git

Aksima commented:
Sorry, I thought I was pushing to the fork? I am very unexperienced with GitHub, so please forgive me.

glennj commented:
You did push to your fork, that's fine. You just opened the pull request before it was ready.

You're missing:
* `gradlew`, `gradlew.bat`, `gradle_wrapper/` -- copy these from another exercise
* the test class `src/test/groovy/LargestSeriesProductSpec.groovy`
* the stub solution class `src/main/groovy/LargestSeriesProduct.groovy`
* the reference solution `.meta/src/reference/groovy/LargestSeriesProduct.groovy`

Aksima commented:
Hello, Glenn Jackman!

I did some work on missing files, here is the report:

## A work on Gradle

### Switch to building directory

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~
$ cd ~/exercism/building-exercism
```

### Clone the repository, as scary error message forced me to delete it

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism
$ git clone --branch largest-series-product --single-branch https://github.com/Aksima/groovy.git
Cloning into 'groovy'...
```

### Move to sample exercise directory

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism
$ cd groovy/exercises/practice/hello-world
```

### Copy gradle files

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/hello-world (largest-series-product)
$ find -name '*gradle*' -exec cp -r {} ../largest-series-product \;
```

### Remove excess files

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/hello-world (largest-series-product)
$ cd ../largest-series-product

Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ find gradle-wrapper.* -exec rm {} \;
```

## A work on test class

### Copy test class file

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ cp ../hello-world/src/test/groovy/HelloWorldSpec.groovy src/test/groovy/LargestSeriesProductSpec.groovy
```

### Correct the test class name

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ sed -i 's/HelloWorld/LargestSeriesProduct/' src/test/groovy/LargestSeriesProductSpec.groovy
```

## A work on stub solution

### Copy stub solution file

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ cp ../hello-world/src/main/groovy/HelloWorld.groovy src/main/groovy/LargestSeriesProduct.groovy
```

### Correct the class name

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ sed -i 's/HelloWorld/LargestSeriesProduct/' src/main/groovy/LargestSeriesProduct.groovy
```

## A work on reference solution

### Copy reference solution file

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ cp ../hello-world/.meta/src/reference/groovy/HelloWorld.groovy .meta/src/reference/groovy/LargestSeriesProduct.groovy
```

### Correct the class name

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ sed -i 's/HelloWorld/LargestSeriesProduct/' .meta/src/reference/groovy/LargestSeriesProduct.groovy
```

## Commiting the solution

### Adding staged files.

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ git add .
```

### Commiting the solution

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ git commit -m 'Copied gradle and groovy files from hello-world exercise and updated class names'
[largest-series-product 4da7f26] Copied gradle and groovy files from hello-world exercise and updated class names
 8 files changed, 336 insertions(+)
```

### Pushing solution to remote

```
Aksima@DESKTOP-NIMQE87 MINGW64 ~/exercism/building-exercism/groovy/exercises/practice/largest-series-product (largest-series-product)
$ git push

glennj commented:
Good!

Next steps: 

* run `bin/configlet lint` and fix the JSON error in config.json.
* implement the tests listed in tests.toml

BNAndras commented on exercises/practice/largest-series-product/.meta/config.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "authors": [],
```
You should put your Github username here as the exercise author to get authoring credit / extra rep on the Exercism website.

BNAndras commented on exercises/practice/largest-series-product/src/test/groovy/LargestSeriesProductSpec.groovy:
```diff
@@ -0,0 +1,212 @@
+import spock.lang.*
+
+class ExerciseSpec extends Specification {
+    @IgnoreRest
```
Exercism prefers tests be enabled one at a time in a TDD fashion. Can we do @Ignore for each of the subsequent tests instead of @IgnoreRest here? That’ll match what has been done on other test suites on this track.

Aksima replied:
Added an ability to generate `@Ignore` attributes to test generator.

BNAndras commented on exercises/practice/largest-series-product/src/test/groovy/LargestSeriesProductSpec.groovy:
```diff
@@ -0,0 +1,212 @@
+    @IgnoreRest
+    def "finds the largest product if span equals length"() {
+        expect:
+        // Please check the "largestProduct" property
```
We probably should remove the please check, expect, and use comments. They’re apparently intended for the test suite author, not the student. The comments don’t provide information that isn’t in the test case itself so they’re noisy.

Aksima replied:
Updated test generator in such a way that it can generate all the code, not just stubs with comments. Comments are not generated anymore.

glennj commented on exercises/practice/largest-series-product/src/test/groovy/LargestSeriesProductSpec.groovy:
```diff
@@ -0,0 +1,182 @@
+        "1234a5" | 2
+    }
+
+//    @Ignore
+//    def "rejects negative span"() {
+//        when:
+//        LargestSeriesProduct.largestProduct(digits, span)
+//
+//        then:
+//        IllegalArgumentException exceptionThrown = thrown(IllegalArgumentException)
+//        exceptionThrown.message == "span must be greater than zero"
+//
+//        where:
+//        digits  | span
+//        "12345" | -1
+//    }
```
Don't keep this: students will get confused when one test is commented out.

glennj commented:
Please run `bin/configlet fmt -e largest-series-product -u`

Aksima commented:
Removed commented test and formatted config.json.

glennj commented:
@Aksima thanks for your hard work to prepare this PR.

@BNAndras care to review?

BNAndras commented on exercises/practice/largest-series-product/src/test/groovy/LargestSeriesProductSpec.groovy:
```diff
@@ -0,0 +1,168 @@
+
+        then:
+        IllegalArgumentException exceptionThrown = thrown(IllegalArgumentException)
+        exceptionThrown.message == "digits input must only contain digits"
```
If we're expecting an argument with a particular message, I feel that message should go in the where block as our `expected` value. That keeps consistency with the other test cases which have `digits`, `span`, and `expected`.

Aksima replied:
@BNAndras, thanks for review!

Unified the "where:" blocks in https://github.com/exercism/groovy/pull/437/commits/9c0889c515a813203e27e5f7e4dd0e1f282dcd45

glennj commented:
@Aksima thank you for getting this done!
